Insert end_cow_mutation_addr for lifetime dependent values dependent on mutable addresses

meg-gupta opened the pull request:
Array/ArraySlice/ContiguousArray have support for mutableSpan property. These types are "optimized COW" types, we use compiler builtins begin_cow_mutation/end_cow_mutation to optimize uniqueness checks. Since mutableSpan is a property and not a coroutine there is no way to schedule the end_cow_mutaton operation at the end of the access.
This can lead to miscompiles in rare cases where we can end up using a stale storage buffer after a cow.
    
This PR inserts end_cow_mutation_addr to avoid this issue.
    
Note: We can end up with unnecessary end_cow_mutation_addr. But it is just a barrier to prevent invalid optimizations
and has no impact.

Fixes rdar://146785284

meg-gupta committed:
Disable runtime cow verification for mutableSpan property

meg-gupta committed:
Introduce end_cow_mutation_addr instruction

meg-gupta committed:
Bridge getSwiftMutableSpanDecl() and isBuiltinType()

eeckstein commented on stdlib/public/core/Array.swift:
```diff
@@ -1735,10 +1746,12 @@ extension Array {
     @lifetime(&self)
     @_alwaysEmitIntoClient
     mutating get {
+#if INTERNAL_CHECKS_ENABLED && COW_CHECKS_ENABLED
```
This doesn't make sense. If checks are enabled then you call the unchecked version?
This completely disables the checks in all cases.
Is this intention?

meg-gupta replied:
Yes. This is disabling runtime verification for mutableSpan property completely. I will remove `#else` parts to remove the confusion.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LifetimeDependenceScopeFixup.swift:
```diff
@@ -121,11 +121,84 @@ let lifetimeDependenceScopeFixupPass = FunctionPass(
+  }
+
+  for endInstruction in scoped!.endInstructions {
+    let builder = Builder(before: endInstruction, location: endInstruction.location, context)
```
You don't need to add the `location:` argument. If not present the builder picks the location from the insertion point anyway

atrick commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LifetimeDependenceScopeFixup.swift:
```diff
@@ -121,11 +121,84 @@ let lifetimeDependenceScopeFixupPass = FunctionPass(
+        return true
+      }
+      return cases.contains { $0.payload?.mayHaveMutableSpan(function: function, context) ?? false }
+    }
```
This check is only looking for the MutableSpan returned by Array.mutableSpan. Storing that into a class would already have escaped it. The class itself if Escapable, so not interesting.

meg-gupta commented on docs/SIL/Instructions.md:
````diff
@@ -2146,6 +2146,19 @@ not replace this reference with a not uniquely reference object.
+end_cow_mutation_addr %0 : $*T
+// %0 must be of an address $*T type
+```
+
````
Updated to:

```
sil-instruction ::= 'end_cow_mutation_addr' sil-operand

end_cow_mutation_addr %0 : $*T
// %0 must be of an address $*T type
```

This instruction marks the end of mutation of an address that may contain MutableSpan.
The address could be an opaque archetype, a struct type, tuple type or enum type and
the end_cow_mutation_addr will apply to all members contained within these types. 
It is currently only generated in cases where is it is not possible to schedule an
`end_cow_mutation` in the standard library automatically. Ex: Array.mutableSpan

meg-gupta commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LifetimeDependenceScopeFixup.swift:
```diff
@@ -121,11 +121,84 @@ let lifetimeDependenceScopeFixupPass = FunctionPass(
+    if isBuiltinType {
+      return false
+    }
+    if nominal == nil {
```
Thanks. This will apply to nominal types only.

atrick reviewed (approved):
Thanks! Only a few comments.

meg-gupta committed:
Insert end_cow_mutation_addr for lifetime dependent values dependent on mutable addresses

Array/ArraySlice/ContiguousArray have support for mutableSpan property.
These types are "optimized COW" types, we use compiler builtins begin_cow_mutation/end_cow_mutation to
optimize uniqueness checks. Since mutableSpan is a property and not a coroutine there is
no way to schedule the end_cow_mutaton operation at the end of the access.
This can lead to miscompiles in rare cases where we can end up using a stale storage buffer after a cow.

This PR inserts end_cow_mutation_addr to avoid this issue.

Note: We can end up with unnecessary end_cow_mutation_addr. But it is just a barrier to prevent invalid optimizations
and has no impact.

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test
